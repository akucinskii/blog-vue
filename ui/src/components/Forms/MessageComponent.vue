<script setup lang="ts">
import { computed } from "vue";
const props = defineProps({
  text: {
    type: String,
    required: false,
    default: "",
  },
  type: {
    type: String,
    required: false,
    default: "success",
  },
});

const type = computed(() => {
  if (props.type === "error") {
    return "error";
  } else {
    return "success";
  }
});
</script>

<template>
  <v-row class="my-4">
    <v-alert id="message" :type="type" :color="type" icon="mdi-alert-circle-outline" close text-color="white" my="2">
      {{ props.text }}
    </v-alert>
  </v-row>
</template>
