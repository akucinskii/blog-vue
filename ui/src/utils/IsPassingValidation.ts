import { RuleI } from "../types/types";

export const isPassingValidation = (tested: string, rules: RuleI[]) =>
  rules.every((rule: RuleI) => rule(tested) === true);
