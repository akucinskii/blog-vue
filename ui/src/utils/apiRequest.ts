import axios from "axios";

const BASE_API_URL = process.env.API_URL;

export enum HTTP_METHODS {
  GET = "GET",
  POST = "POST",
  PUT = "PUT",
  DELETE = "DELETE",
  PATCH = "PATCH",
}

axios.defaults.baseURL = process.env.API_URL;

const apiRequest = async (method: HTTP_METHODS, url: string, data = {}, headers = {}) => {
  const token = localStorage.getItem("token");

  axios.defaults.headers.common["Authorization"] = "Bearer " + token;

  const response = await axios({
    method,
    url: `${BASE_API_URL}${url}`,
    data,
    headers,
  });

  return response;
};

export default apiRequest;
