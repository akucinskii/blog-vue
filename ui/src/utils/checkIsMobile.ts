const checkIsMobile = (): boolean => {
  return window.innerWidth < 1264;
};

export default checkIsMobile;
