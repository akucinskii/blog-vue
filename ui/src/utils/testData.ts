const testPosts = [
  {
    id: "313553ff-784f-497f-adfc-7e549bebab66",
    title:
      "POST TESTOWY asm.nfnasnfas.fa.sfn,acsaf;;;;;;;;;;;;jslfsafsjafjsajkfsjfsajfsafsajlfjlksajklfasjklafsjklfsajklaflkjslksfnsaf,mnfsa,fsannm",
    date: "2037-01-01",
    text: "It is a long established fact that a reader will be distracted by the readable content of a page when looking at its layout. The point of using Lorem Ipsum is that it has a more-or-less normal distribution of letters, as opposed to using 'Content here, content here', making it look like readable English. Many desktop publishing packages and web page editors now use Lorem Ipsum as their default model text, and a search for 'lorem ipsum' will uncover many web sites still in their infancy. Various versions have evolved over the years, sometimes by accident, sometimes on purpose (injected humour and the like).",
    imageUrl:
      "https://images.unsplash.com/photo-1495753379358-73c76ccd644b?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1539&q=80",
    author: {
      username: "Robert",
      avatar:
        "https://images.unsplash.com/photo-1495753379358-73c76ccd644b?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1539&q=80",
      description: "test description",
    },
    category: {
      id: "1",
      name: "Technology",
    },
    disabled: false,
  },
  {
    id: "313553ff-784f-497f-adfc-7e549rerab66",
    title: "sfakfasjlasdasdasndaskdjaskjsj",
    date: "2020-01-01",
    text: "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi e",
    imageUrl:
      "https://images.unsplash.com/photo-1495753379358-73c76ccd644b?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1539&q=80",
    author: {
      username: "Wojciech",
      avatar:
        "https://images.unsplash.com/photo-1495753379358-73c76ccd644b?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1539&q=80",
      description: "test description",
    },
    category: {
      id: "2",
      name: "Dieta",
    },
    disabled: false,
  },
  {
    id: "fsafasd",
    title: "Najlepsza dieta w okolicy. Zobacz teraz!!!",
    date: "2020-01-01",
    text: "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi e",
    imageUrl:
      "https://images.unsplash.com/photo-1529565214304-a882ebc5a8e6?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=2070&q=80",
    author: {
      username: "Test author",
      avatar:
        "https://images.unsplash.com/photo-1633621412960-6df85eff8c85?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=627&q=80",
      description: "test description",
    },
    category: {
      id: "1",
      name: "Technology",
    },
    disabled: false,
  },
  {
    id: "fsafasd",
    title: "This post shouldnt be visible",
    date: "2020-01-01",
    text: "This post shouldnt be visible",
    imageUrl:
      "https://images.unsplash.com/photo-1529565214304-a882ebc5a8e6?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=2070&q=80",
    author: {
      username: "Test author",
      avatar:
        "https://images.unsplash.com/photo-1633621412960-6df85eff8c85?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=627&q=80",
      description: "test description",
    },
    category: {
      id: "2",
      name: "Dieta",
    },
    disabled: true,
  },
];

export const testCategories = [
  {
    id: "1",
    name: "Technology",
  },
  {
    id: "2",
    name: "Dieta",
  },
  {
    id: "3",
    name: "Entertainment",
  },
];

export const testUser = {
  id: "1",
  username: "TestUser",
  password: "Password123.",
  avatar_url: "testAvatar.jpg",
  description: "test description",
  name: "Test name",
  surname: "Test surname",
  is_admin: true,
};

export default testPosts;
