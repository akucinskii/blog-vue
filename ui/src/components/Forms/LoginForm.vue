<script setup lang="ts">
import { ref } from "vue";
import { useRouter } from "vue-router";
import NavbarLogo from "../navbar/NavbarLogo.vue";
import { useUserStore } from "../../stores/user";
import { MessageInterface, ResponseInterface } from "../../types/types";
import qs from "qs";
import apiRequest, { HTTP_METHODS } from "../../utils/apiRequest";
import axios, { AxiosError } from "axios";
import MessageComponent from "./MessageComponent.vue";

const username = ref("");
const password = ref("");

const message = ref<MessageInterface>({
  text: "",
  type: "success",
});

const router = useRouter();

const userStore = useUserStore();

const handleLogin = async () => {
  message.value.text = "";

  const headers = {
    "Content-Type": "application/x-www-form-urlencoded",
    accept: "application/json",
  };

  const data = qs.stringify({
    username: username.value,
    password: password.value,
  });

  try {
    const response = await apiRequest(HTTP_METHODS.POST, "/token", data, {
      headers,
    });
    localStorage.setItem("token", response.data.access_token);
    axios.defaults.headers.common["Authorization"] = "Bearer " + localStorage.getItem("token");
    await userStore.fetchUser();
    router.push("/");
  } catch (error) {
    const err = error as AxiosError<ResponseInterface>;

    message.value.type = "error";

    switch (err.response?.status) {
      case 422:
        message.value.text = "Fields cannot be empty";
        break;
      default:
        message.value.text = err.response?.data.detail ?? "Something went wrong";
        break;
    }
  }
};

defineExpose({
  username,
  password,
  message,
  handleLogin,
});
</script>

<template>
  <v-sheet elevation="23" rounded="lg" class="pa-8">
    <v-row justify="center" class="my-4">
      <NavbarLogo />
    </v-row>
    <MessageComponent v-if="message.text" :type="message.type" :text="message.text" />
    <v-form @submit.prevent="handleLogin">
      <v-text-field
        id="usernameInput"
        v-model="username"
        prepend-outer-icon="mdi-account"
        label="username"
        variant="underlined"
        type="username"
        required
        autofocus
      ></v-text-field>
      <v-text-field
        id="passwordInput"
        v-model="password"
        prepend-outer-icon="mdi-lock"
        label="Password"
        variant="underlined"
        type="password"
        required
      ></v-text-field>
      <v-row justify="center" class="mb-4">
        <v-btn id="submitButton" color="red" class="text--white px-8" type="submit">Login</v-btn>
      </v-row>
      <v-row justify="center">
        <router-link to="/register"><v-btn variant="text" class="text--gray"> Register </v-btn></router-link>
      </v-row>
    </v-form>
  </v-sheet>
</template>

<style lang="scss" scoped>
.text--white {
  color: white !important;
}
</style>
