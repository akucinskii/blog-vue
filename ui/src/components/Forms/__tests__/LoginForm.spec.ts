import { createTestingPinia } from "@pinia/testing";
import { mount, shallowMount } from "@vue/test-utils";
import { AxiosResponse } from "axios";
import { mocked } from "ts-jest/utils";
import { nextTick } from "vue";
import apiRequest from "../../../utils/apiRequest";
import LoginForm from "../LoginForm.vue";
import MessageComponent from "../MessageComponent.vue";

jest.mock("../../../utils/apiRequest");
const mockedApiRequest = mocked(apiRequest);

describe("LoginForm", () => {
  test("renders without crashing", () => {
    const wrapper = mount(LoginForm, {
      global: {
        plugins: [createTestingPinia()],
      },
    });
    expect(wrapper.exists()).toBe(true);
  });

  test("Matches snapshot", () => {
    const wrapper = mount(LoginForm, {
      global: {
        plugins: [createTestingPinia()],
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  test("Form is empty when mounted", () => {
    const wrapper = mount(LoginForm, {
      global: {
        plugins: [createTestingPinia()],
      },
    });

    expect(wrapper.vm.password).toEqual("");
    expect(wrapper.vm.username).toEqual("");
  });

  test("api request is made when button is clicked", async () => {
    const wrapper = shallowMount(LoginForm, {
      global: {
        plugins: [createTestingPinia()],
      },
    });

    wrapper.vm.username = "test";
    wrapper.vm.password = "test";

    mockedApiRequest.mockResolvedValueOnce({
      data: {
        token: "token",
      },
    } as AxiosResponse);

    await wrapper.find("v-form").trigger("submit.prevent");
    await nextTick();

    expect(mockedApiRequest).toHaveBeenCalled();
  });

  test("displays correct error message when there is an error 422", async () => {
    const wrapper = mount(LoginForm, {
      global: {
        plugins: [createTestingPinia()],
      },
    });

    wrapper.vm.username = "test";
    wrapper.vm.password = "test";

    mockedApiRequest.mockRejectedValueOnce({
      response: {
        status: 422,
        data: {
          detail: "error",
        },
      },
    });

    await wrapper.find("v-form").trigger("submit.prevent");
    await nextTick();

    expect(wrapper.findComponent(MessageComponent).text()).toContain("Fields cannot be empty");
  });
  test("displays correct error message when there is an other error", async () => {
    const wrapper = mount(LoginForm, {
      global: {
        plugins: [createTestingPinia()],
      },
    });

    wrapper.vm.username = "test";
    wrapper.vm.password = "test";

    mockedApiRequest.mockRejectedValueOnce({
      response: {
        status: 404,
        data: {
          detail: "error other than 422",
        },
      },
    });

    await wrapper.find("v-form").trigger("submit.prevent");
    await nextTick();

    expect(wrapper.findComponent(MessageComponent).text()).toContain("error other than 422");
  });
});
