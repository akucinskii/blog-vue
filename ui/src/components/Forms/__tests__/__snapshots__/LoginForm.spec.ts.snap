// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm Matches snapshot 1`] = `
<v-sheet
  class="pa-8"
  elevation="23"
  rounded="lg"
>
  <v-row
    class="my-4"
    justify="center"
  >
    <router-link
      to="/"
    >
      Dieta by Robercik
    </router-link>
  </v-row>
  <!--v-if-->
  <v-form>
    <v-text-field
      autofocus=""
      id="usernameInput"
      label="username"
      modelvalue=""
      prepend-outer-icon="mdi-account"
      required=""
      type="username"
      variant="underlined"
    />
    <v-text-field
      id="passwordInput"
      label="Password"
      modelvalue=""
      prepend-outer-icon="mdi-lock"
      required=""
      type="password"
      variant="underlined"
    />
    <v-row
      class="mb-4"
      justify="center"
    >
      <v-btn
        class="text--white px-8"
        color="red"
        id="submitButton"
        type="submit"
      >
        Login
      </v-btn>
    </v-row>
    <v-row
      justify="center"
    >
      <router-link
        to="/register"
      >
        <v-btn
          class="text--gray"
          variant="text"
        >
           Register 
        </v-btn>
      </router-link>
    </v-row>
  </v-form>
</v-sheet>
`;
